clang/lib/CodeGen/CGExprScalar.cpp

```diff
@@ -2249,19 +2249,17 @@ static void warnAboutImplicitCToPtr(CodeGenModule &CGM, CastExpr *CE) {
 }
 
 static llvm::Value *createCToPtr(CodeGenFunction &CGF, llvm::Value *Cap,
-                                 llvm::Type *LLVMTy, QualType Ty) {
+                                 llvm::Type *LLVMTy) {
   assert(llvm::isCheriPointer(Cap->getType(), &CGF.CGM.getDataLayout()));
   assert(LLVMTy->isIntegerTy());
   auto DDC =
       CGF.Builder.CreateIntrinsic(llvm::Intrinsic::cheri_ddc_get, {}, {});
   auto CToPtr = CGF.Builder.CreateIntrinsic(
       llvm::Intrinsic::cheri_cap_to_pointer, CGF.PtrDiffTy,
       {DDC, CGF.Builder.CreatePointerCast(Cap, CGF.Int8CheriCapTy)});
-  // sign extend if the target is a signed integer type otherwise zext.
-  // This should only happen when assigning the result to a larger type:
-  // i.e. __int128 or int64_t for CHERI64
-  return CGF.Builder.CreateIntCast(CToPtr, LLVMTy,
-                                   Ty->isSignedIntegerOrEnumerationType());
+  // Truncate or zero-extend if the result type is not the same width as the
+  // address type to match ptrtoint's semantics.
+  return CGF.Builder.CreateIntCast(CToPtr, LLVMTy, /*isSigned=*/false);
```

Comment:
This appears to be untested since there is no test diff?

Right now the following performs a sext, but now it will be a zext. Could you add a test?
```
__int128 test(void* __capability p) {
    return (__int128)p;
}
```

https://cheri-compiler-explorer.cl.cam.ac.uk/z/b48461

Reply:
That's what cap_to_oversized is?

Reply:
It is indeed - not sure how I didn't see that...